### ISSUE TITLE
Long-lived global connectors become eventually unresponsive

### ISSUE THREAD

**Issue by SvenHoeffler** [openintegrationhub/openintegrationhub#1139]
When a global connector is left active for a prolonged time, it may end up becoming unable to emit new events. Instead, it logs an error along the lines of `AMQP buffer full, awaiting drain event...`. This effectively stops that connector from functioning, blocking any further execution of flows involving this connector.

Restarting the affected global connector through the Component Orchestrator appears to fix this.

---

[**Definition of Done**](https://github.com/openintegrationhub/openintegrationhub/blob/crudmonitoring/CONTRIBUTING.md#definition-of-done)

**Comment by SvenHoeffler** [comment]
Update: This issue can also affect connectors that have only recently been started. We've had a connector affected by it only about half an hour after it had been started. Curiously, the issue was only resolved after *all other* connectors had been restarted.

This suggests that the issue is somewhere within the queue itself, and possibly caused by one or several connectors eventually overwhelming it in some way.

**Comment by weberjm** [comment]
To provide more context for future review:

The error which is mentioned seems to be the one generated from https://github.com/openintegrationhub/openintegrationhub/blob/97fe6cbe3c978697d27def782c5db687ee096089/lib/ferryman/lib/amqp.js#L217-L221
in the AMQP class of the Ferryman library. Once this error message is published, the component is set to await the 'drain' event from the publishChannel, which is an AMQP `confirmChannel`. 

For some reason, the drain event never seems to be emitted, so the component continues waiting. To investigate: what is the rabbitMQ drain event behavior? Is the behavior different on confirm channels rather than standard?

**Comment by weberjm** [comment]
_Side Note:_ The package `event-to-promise` which is used in the above snippet is deprecated and has not had an update in 4 years. They have a recommended replacement on the package page in npm. If work needs to be done on the amqp files, we should consider replacing that.

**Comment by weberjm** [comment]
After some further investigation and consultation (thanks @SvenHoeffler!), it is likely that this section of code could in certain cases be causing the component to incorrectly believe that the buffer is full, and then block waiting for a `drain` event which will never be emitted. https://github.com/openintegrationhub/openintegrationhub/blob/3ef80d9de5e28898feda1df44295462f05ac2157/lib/ferryman/lib/amqp.js#L204-L230

When the Promise.all puts both the async function and the publishPromise, the publishPromise appears to always execute first *if* the underlying code is completely synchronous (I have tested in a REPL, and been able to obtain results confirming this, although I am not 100% sure on the task and microtask queue behavior for a Promise.all() call. I believe that the publishPromise is added to the task queue first, while the added layer of the async function call puts its Promise execution later in the task queue, thus allowing the Promise resolution to always end up in the microtask queue before the async function makes it to the call stack.) In this way, the asynchronous function always has the value of result to use in resolving its promise.

I believe that the problem arises if amqp then makes an asynchronous call, which would then delay the resolution of the publishPromise and allow the async function to begin execution without a result available. It then assumes the buffer is full, and then waits for the 'drain' event which would never come.

(Note: in reading documentation about the amqplib library, this line stuck out to me:
> Amqplib exposes synchronous and asynchronous methods over the same API.

and there could be internal implementations that are sometimes synchronous and sometimes asynchronous, leading to this behavior. 

A potential first step in analysing this problem could be to remove the Promise.all() from the function, wait on the result of publishPromise, and then do the buffer full check.

**Comment by SvenHoeffler** [comment]
A fix has just been merged in. The two promises are now strictly sequential, ensuring that the correct result of `publish()` is known before checking for the drain clause. 

Additionally, we've introduced a timeout to ensure that the process can no longer be indefinitely stuck in this clause waiting for a drain event that never comes. For now, the timeout threshold is set to 10 seconds, which ought to be plenty, but we may want to keep an eye on this and adjust it as needed.
